expr/func_linearregression.go

```diff
@@ -0,0 +1,166 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/schema"
+	"github.com/raintank/dur"
+)
+
+type FuncLinearRegression struct {
+	in GraphiteFunc
+
+	startSourceAt string // at(1) format
+	endSourceAt   string
+	startSource   uint32 // epoch seconds
+	endSource     uint32
+
+	startTarget uint32 // epoch seconds
+	endTarget   uint32
+}
+
+func NewLinearRegression() GraphiteFunc {
+	return &FuncLinearRegression{}
+}
+
+func (s *FuncLinearRegression) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{
+				val: &s.in,
+			},
+			ArgString{
+				key:       "startSourceAt",
+				opt:       true,
+				validator: []Validator{IsRenderTimeFormat},
+				val:       &s.startSourceAt,
+			},
+			ArgString{
+				key:       "endSourceAt",
+				opt:       true,
+				validator: []Validator{IsRenderTimeFormat},
+				val:       &s.endSourceAt,
+			},
+		},
+		[]Arg{ArgSeriesList{}}
+}
+
+func (s *FuncLinearRegression) Context(context Context) Context {
+	s.startTarget = context.from
+	s.endTarget = context.to
+
+	if err := s.parseSourceAt(); err != nil {
+		return context // todo panic?
```

Comment:
I'm not sure of the correct way to handle this failure. Some functions seem to just log and continue.